extensions/interactions/SetInput/directives/set-input-validation.service.spec.ts

```diff
@@ -74,145 +72,731 @@ describe('SetInputValidationService', () => {
+              x: ['3', '4', '5', '6']
+            }
+          });
+          let answerGroup = createAnswerGroupByRules(
+            [equalsRule1, equalsRule2]
+          );
+
+          let warnings = validatorService.getAllWarnings(
+            currentState,
+            goodCustomizationArgs,
+            [answerGroup],
+            goodDefaultOutcome
+          );
+          expect(warnings).toEqual([]);
+        });
+      });
+
+      describe('IsSubsetOf', () => {
+        it('should generate errors with identical rules', () => {
+          let subsetRule = rof.createFromBackendDict({
+            rule_type: 'IsSubsetOf',
+            inputs: {
+              x: ['1', '2', '3']
+            }
+          });
+          let answerGroup = createAnswerGroupByRules(
+            [subsetRule, subsetRule]
+          );
+
+          let warnings = validatorService.getAllWarnings(
+            currentState,
+            goodCustomizationArgs,
+            [answerGroup],
+            goodDefaultOutcome
+          );
+          expect(warnings).toEqual([{
+            type: WARNING_TYPES.ERROR,
+            message: 'Rule 2 from answer group 1 is the same as rule 1 ' +
+            'from answer group 1'
+          }]);
+        });
+
+        it('should not generate errors with different rules', () => {
+          let subsetRule1 = rof.createFromBackendDict({
+            rule_type: 'IsSubsetOf',
+            inputs: {
+              x: ['1', '2', '3']
+            }
+          });
+          let subsetRule2 = rof.createFromBackendDict({
+            rule_type: 'IsSubsetOf',
+            inputs: {
+              x: ['3', '4', '5', '6']
+            }
+          });
+          let answerGroup = createAnswerGroupByRules(
+            [subsetRule1, subsetRule2]
+          );
+
+          let warnings = validatorService.getAllWarnings(
+            currentState,
+            goodCustomizationArgs,
+            [answerGroup],
+            goodDefaultOutcome
+          );
+          expect(warnings).toEqual([]);
+        });
+      });
+
+      describe('HasElementsIn', () => {
+        it('should generate errors with identical rules', () => {
+          let hasElementsInRule = rof.createFromBackendDict({
+            rule_type: 'HasElementsIn',
+            inputs: {
+              x: ['1', '2', '3']
+            }
+          });
+          let answerGroup = createAnswerGroupByRules(
+            [hasElementsInRule, hasElementsInRule]
+          );
+
+          let warnings = validatorService.getAllWarnings(
+            currentState,
+            goodCustomizationArgs,
+            [answerGroup],
+            goodDefaultOutcome
+          );
+          expect(warnings).toEqual([{
+            type: WARNING_TYPES.ERROR,
+            message: 'Rule 2 from answer group 1 is the same as rule 1 ' +
+            'from answer group 1'
+          }]);
+        });
+
+        it('should not generate errors with different rules', () => {
```

Comment:
Aren't these the same rules with different inputs? Ditto elsewhere (as applicable).

Reply:
Done.